src/QuadGK.jl

```diff
@@ -140,14 +123,14 @@ function do_quadgk(f, s, n, ::Type{Tw}, abstol, reltol, maxevals, nrm) where Tw
     # Pop the biggest-error segment and subdivide (h-adaptation)
     # until convergence is achieved or maxevals is exceeded.
     while E > abstol && E > reltol * nrm(I) && numevals < maxevals
-        s = heappop!(segs, Reverse)
-        mid = (s.a + s.b) * 0.5
-        s1 = evalrule(f, s.a, mid, x,w,gw, nrm)
-        s2 = evalrule(f, mid, s.b, x,w,gw, nrm)
-        heappush!(segs, s1, Reverse)
-        heappush!(segs, s2, Reverse)
-        I = (I - s.I) + s1.I + s2.I
-        E = (E - s.E) + s1.E + s2.E
+        seg = heappop!(segs, Reverse)
+        mid = (seg.a + seg.b) * 0.5
+        seg1 = evalrule(f, seg.a, mid, x,w,gw, nrm)
+        seg2 = evalrule(f, mid, seg.b, x,w,gw, nrm)
+        heappush!(segs, seg1, Reverse)
```

Comment:
Seems to need some kind of check here for `seg1 isa eltype(segs)` (similarly for `seg2`), and reallocate `segs` if not (will also require a refactored `do_quadgk_common` routine that can take an existing `segs` heap as a starting point).

Reply:
this would mean all segments don't have the same type? Wouldn't that be strange, boundaries all have the same type?